DOCS-4086: code samples for building a good dataset (#4413)
viamrobotics/docs: docs/data-ai/train/create-dataset.md

Reviewer: I think adding all the sdk tabs, and adding the cards, helped bulk this page out a little more so it's not as terribly small. anymore. Still small, but I guess with the tabs there's enough content that it doesn't fit anywhere else nicely.

Reviewer: Right now there is still capture and add to dataset content on two different pages which feels confusing.
It's hard to force a linear flow here since with all the different options for each step, the same order doesn't always apply:
- In most cases, like using the mobile app, uploading a batch, or adding existing images, you _can_ get the data first and then create a dataset and add the data to it. Order doesn't matter, but it's easiest conceptually to think of getting data and then making a dataset with it.
- One exception: If you [capture individual images](https://deploy-preview-4413--viam-docs.netlify.app/data-ai/train/capture-annotate-images/#capture-individual-images) **through the UI and add them to a dataset on the spot**, you _have to have a dataset_ to add them to before you start capturing.
- In the script version of that same "Capture individual images" heading, it looks like you don't specify a dataset id, so you'd still have to add to a dataset later like with the other methods.

Suggestion:
- Implement Naomi's order suggestion, and:
- Get rid of the exception: Ask eng to change that capture button to *not* save to dataset but rather just save the image to your captured data. Just one single click, so you can capture more images in rapid succession, then add a batch all at once later. This would be a less clunky UX IMO, and also solve a docs flow problem.
   - If this will take a while but they'll do it, don't worry about this flow for now; document the rest per Naomi's order
- If this can't/won't ever be changed in eng, document this as a thing you can do but shape the docs around the normal capture-then-make-a-dataset order

Reviewer 2: but overall I do like this change a lot - thank you for making it. Regardless of which order. Just need to make sure the copy references things in the correct order

const client = await createViamClient();
const dataClient = client.dataClient;

console.log("Creating dataset...");

try {
  const datasetId = await dataClient.createDataset({
    name: "<dataset_name>",
    organizationId: "<org_id>",
  });
  console.log(`Created dataset: ${datasetId}`);
} catch (error) {
  console.log("Error creating dataset. It may already exist.");
  console.log(`Exception: ${error}`);
  process.exit(1);
}
```

{{% /tab %}}
{{% tab name="Flutter" %}}

To create a dataset, pass a unique dataset name and organization ID to [`dataClient.createDataset`](/dev/reference/apis/data-client/#createdataset):

```dart
final viamClient = await ViamClient.connect();
final dataClient = viamClient.dataClient;

print("Creating dataset...");

try {
    final datasetId = await dataClient.createDataset(
        name: "<dataset_name>",
        organizationId: "<org_id>",
    );
    print("Created dataset: $datasetId");
} catch (e) {
    print("Error creating dataset. It may already exist.");
    print("Exception: $e");
    return;
}
```

{{% /tab %}}
{{< /tabs >}}

Finish creating a dataset by adding annotated images to it.
You can capture new images or add existing images:

> ## Add to a dataset

Reviewer: If the next heading is "Add all images captured by a specific machine to a dataset," What specifically does just "Add to a dataset" mean? Maybe make this heading more descriptive and also add an intro sentence

Note this comment may be moot when/if Naomi's suggestion to do 1. Capture + annotate then 2. Create a dataset is applied

```typescript
const client: ViamClient = await createViamClient({
  credential: {
    type: "api-key",
    payload: API_KEY,
  },
  authEntity: API_KEY_ID,
});

const dataClient = client.dataClient;

// Add image to dataset
await dataClient.addBinaryDataToDatasetByIds({
  binaryIds: [EXISTING_IMAGE_ID],
  datasetId: EXISTING_DATASET_ID,
});

console.log("Image added to dataset successfully");
client.disconnect();
```

{{% /tab %}}
{{% tab name="Flutter" %}}

To add an image to a dataset, find the binary data ID for the image and the dataset ID.
Pass both IDs to [`dataClient.addBinaryDataToDatasetByIDs`](/dev/reference/apis/data-client/#addbinarydatatodatasetbyids):

```dart
final client = await ViamClient.withApiKey(
    apiKeyId: apiKeyId,
    apiKey: apiKey,
);

final dataClient = client.dataClient;

try {
    // Add image to dataset
    await dataClient.addBinaryDataToDatasetByIds(
      binaryIds: [existingImageId],
      datasetId: existingDatasetId,
    );

    print('Image added to dataset successfully');
} finally {
    await client.close();
}
```

{{% /tab %}}
{{< /tabs >}}

## Add all images captured by a specific machine to a dataset

> The following script adds all images captured from a certain machine to a new dataset:

Reviewer: The first tab shows the UI, so this text should go on each SDK tab instead
```suggestion
```

{{< tabs >}}
{{% tab name="Web UI" %}}

You can add images to a dataset from the **Images** tab of the [**DATA** page](https://app.viam.com/data/view):

1. From the **Machine name** dropdown, select the name of a machine.
1. Click the **Apply** button at the bottom of the left sidebar.
1. Click to select the images you would like to add to your dataset.
1. Click the **Add to dataset** button in the top right.
1. From the **Dataset** dropdown, select the name of your dataset.
1. Click **Add \<n\> images** to add the selected images to the dataset.

{{< alert title="Tip" color="tip" >}}

To select a range of images, select one image, then hold **Ctrl/Cmd** while clicking another image.
This will select both images as well as the entire range of images between those images.

{{< /alert >}}

{{% /tab %}}
{{% tab name="Python" %}}

> 

Reviewer: ```suggestion

The following script adds all images captured from a certain machine to a new dataset:

```

            break
        all_matches.extend(data)

    return all_matches

async def main() -> int:
    """Main execution function."""
    viam_client = await connect()
    data_client = viam_client.data_client

    matching_data = await fetch_binary_data_ids(data_client, PART_ID)

    print("Creating dataset...")

    try:
        dataset_id = await data_client.create_dataset(
            name=DATASET_NAME,
            organization_id=ORG_ID,
        )
        print(f"Created dataset: {dataset_id}")
    except Exception as e:
        print("Error creating dataset. It may already exist.")
        print("See: https://app.viam.com/data/datasets")
        print(f"Exception: {e}")
        return 1

    print("Adding data to dataset...")

    await data_client.add_binary_data_to_dataset_by_ids(
        binary_ids=[
            obj.metadata.binary_data_id for obj in matching_data
        ],
        dataset_id=dataset_id
    )

    print("Added files to dataset.")
    print(
        f"See dataset: https://app.viam.com/data/datasets?id={dataset_id}"
    )

    viam_client.close()
    return 0

if __name__ == "__main__":
    asyncio.run(main())
```

{{% /tab %}}
{{% tab name="Go" %}}

> 

Reviewer: ```suggestion

The following script adds all images captured from a certain machine to a new dataset:

```

    if err != nil {
        log.Fatalf("Failed to fetch binary data: %v", err)
    }

    fmt.Println("Creating dataset...")

    datasetResp, err := dataClient.CreateDataset(ctx, &data.CreateDatasetRequest{
        Name:           DATASET_NAME,
        OrganizationId: ORG_ID,
    })
    if err != nil {
        fmt.Println("Error creating dataset. It may already exist.")
        fmt.Println("See: https://app.viam.com/data/datasets")
        fmt.Printf("Exception: %v\n", err)
        return
    }

    datasetID := datasetResp.Id
    fmt.Printf("Created dataset: %s\n", datasetID)

    fmt.Println("Adding data to dataset...")

    var binaryIDs []string
    for _, obj := range matchingData {
        binaryIDs = append(binaryIDs, obj.Metadata.Id)
    }

    _, err = dataClient.AddBinaryDataToDatasetByIds(ctx, &data.AddBinaryDataToDatasetByIdsRequest{
        BinaryIds: binaryIDs,
        DatasetId: datasetID,
    })
    if err != nil {
        log.Fatalf("Failed to add binary data to dataset: %v", err)
    }

    fmt.Println("Added files to dataset.")
    fmt.Printf("See dataset: https://app.viam.com/data/datasets?id=%s\n", datasetID)
}

```

{{% /tab %}}
{{% tab name="TypeScript" %}}

> 

Reviewer: ```suggestion

The following script adds all images captured from a certain machine to a new dataset:

```

    createRequest.setName(DATASET_NAME);
    createRequest.setOrganizationId(ORG_ID);

    const datasetResponse = await dataClient.createDataset(createRequest);
    const datasetId = datasetResponse.getId();
    console.log(`Created dataset: ${datasetId}`);

    console.log("Adding data to dataset...");

    const binaryIds = matchingData.map(
      (obj) => obj.getMetadata()?.getId() || "",
    );

    const addRequest = new AddBinaryDataToDatasetByIdsRequest();
    addRequest.setBinaryIdsList(binaryIds);
    addRequest.setDatasetId(datasetId);

    await dataClient.addBinaryDataToDatasetByIds(addRequest);

    console.log("Added files to dataset.");
    console.log(
      `See dataset: https://app.viam.com/data/datasets?id=${datasetId}`,
    );

    viamClient.disconnect();
    return 0;
  } catch (error) {
    console.log("Error creating dataset. It may already exist.");
    console.log("See: https://app.viam.com/data/datasets");
    console.log(`Exception: ${error}`);
    viamClient.disconnect();
    return 1;
  }
}

if (require.main === module) {
  main().then((code) => process.exit(code));
}
```

{{% /tab %}}
{{% tab name="Flutter" %}}

> 

Reviewer: ```suggestion

The following script adds all images captured from a certain machine to a new dataset:

```

    );

    if (response.data.isEmpty) {
      break;
    }

    allMatches.addAll(response.data);
    last = response.last;
  }

  return allMatches;
}

Future<int> main() async {
  final viamClient = await connect();
  final dataClient = viamClient.dataClient;

  final matchingData = await fetchBinaryDataIds(dataClient, partId);

  print("Creating dataset...");

  try {
    final datasetId = await dataClient.createDataset(
      name: datasetName,
      organizationId: orgId,
    );
    print("Created dataset: $datasetId");

    print("Adding data to dataset...");

    final binaryIds = matchingData
        .map((obj) => obj.metadata.binaryDataId)
        .toList();

    await dataClient.addBinaryDataToDatasetByIds(
      binaryIds: binaryIds,
      datasetId: datasetId,
    );

    print("Added files to dataset.");
    print("See dataset: https://app.viam.com/data/datasets?id=$datasetId");

    viamClient.close();
    return 0;

  } catch (error) {
    print("Error creating dataset. It may already exist.");
    print("See: https://app.viam.com/data/datasets");
    print("Exception: $error");
    viamClient.close();
    return 1;
  }
}
```

{{% /tab %}}
{{< /tabs >}}

> ## Capture, annotate, and add images to a dataset

Reviewer: Depending on where the structure of these pages lands, maybe it'd make sense to frame this more like "Combine these steps in a single script" to acknowledge the overlap with other sections

The following example demonstrates how you can capture an image, use an ML model to generate annotations, and then add the image to a dataset.

> You can use this logic to expand and improve your datasets continuously over time as your application runs.

Reviewer: ```suggestion
You can use this logic to expand and improve your datasets continuously over time.
```
(maybe) I think this implies that the application automatically re-trains the model

> Re-train your ML model on the improved dataset to improve the ML model.

Reviewer: If I'm training the same model upon which I'm relying to make the annotations, with no other checking, how does it improve itself? Won't it just be self-reinforcing in possibly bad ways? Should this explicitly tell users to check the annotations manually before re-training the model?
```suggestion
Check the annotation accuracy in the **DATA** tab, then re-train your ML model on the improved dataset to improve the ML model.
```

---
linkTitle: "Create a training dataset"
title: "Create a training dataset"
weight: 10
layout: "docs"
type: "docs"
description: "Create a dataset to train a machine learning model."
aliases:
  - /fleet/dataset/
  - /manage/data/label/
  - /manage/data/dataset/
  - /data/dataset/
  - /data-ai/ai/create-dataset/
---

To train a machine learning model, you will need a dataset.

You can create a dataset using the web UI, the CLI, or one of the SDKs:

{{< tabs >}}
{{% tab name="Web UI" %}}

1. Navigate to the **DATA** page and open the [**DATASETS** tab](https://app.viam.com/data/datasets).

1. Click the **+ Create dataset** button.

   {{< imgproc src="/services/data/create-dataset.png" alt="The **DATASET** tab of the **DATA** page, showing the **+ Create dataset** button." resize="800x" style="width:500px" class="imgzoom" >}}

1. Enter a unique name for the dataset.

> 1. Click **Create dataset** to create the dataset.

Reviewer: ```suggestion
1. Click **Create dataset**.
```
Mm this feels unnecessary, plus see comment above.

```typescript
const client = await createViamClient();
const dataClient = client.dataClient;

console.log("Creating dataset...");

try {
  const datasetId = await dataClient.createDataset({
    name: "<dataset_name>",
    organizationId: "<org_id>",
  });
  console.log(`Created dataset: ${datasetId}`);
} catch (error) {
  console.log("Error creating dataset. It may already exist.");
  console.log(`Exception: ${error}`);
  process.exit(1);
}
```

{{% /tab %}}
{{% tab name="Flutter" %}}

To create a dataset, pass a unique dataset name and organization ID to [`dataClient.createDataset`](/dev/reference/apis/data-client/#createdataset):

```dart
final viamClient = await ViamClient.connect();
final dataClient = viamClient.dataClient;

print("Creating dataset...");

try {
    final datasetId = await dataClient.createDataset(
        name: "<dataset_name>",
        organizationId: "<org_id>",
    );
    print("Created dataset: $datasetId");
} catch (e) {
    print("Error creating dataset. It may already exist.");
    print("Exception: $e");
    return;
}
```

{{% /tab %}}
{{< /tabs >}}

> Once you create a dataset, [capture](/data-ai/train/capture-images/) images, [add](/data-ai/train/update-dataset/) the images to your dataset, and [annotate](/data-ai/train/annotate-images/) the images with training metadata to train your own ML model.

Reviewer: ```suggestion
Once you create a dataset, [capture](/data-ai/train/capture-images/) images, [add the images](/data-ai/train/update-dataset/) to your dataset, and [annotate](/data-ai/train/annotate-images/) the images with training metadata to train your own ML model.
```
[opt] Would be better for accessibility to have the full "capture images" etc. for each of the three items in the link, for more descriptive link text. But I think you did it this way so that it's visually clear that there are 3 different links here? But, "add" is so tiny that maybe expand that one at least? easier to click on haha. Or better yet just delete these links since you have them below and this is redundant.

> Once you create a dataset, [capture](/data-ai/train/capture-images/) images, [add](/data-ai/train/update-dataset/) the images to your dataset, and [annotate](/data-ai/train/annotate-images/) the images with training metadata to train your own ML model.

Reviewer: Hmm also, "Once you create a dataset" feels strange... why would I capture images *after* creating a dataset? Should this be 
```suggestion
Now that you have a dataset, you can [add](/data-ai/train/update-dataset/) the images to your dataset, and [annotate](/data-ai/train/annotate-images/) the images with training metadata to train your own ML model.

If you need to capture more images for your dataset, see [Capture images for training](/data-ai/train/capture-images/).
```

Hmmm looking into this more I now see we present it in different orders in different places.... will comment elsewhere as well.

> Once you create a dataset, [capture](/data-ai/train/capture-images/) images, [add](/data-ai/train/update-dataset/) the images to your dataset, and [annotate](/data-ai/train/annotate-images/) the images with training metadata to train your own ML model.

Reviewer: Okay final answer :D How about something like this?
```suggestion
Finish creating a dataset by adding annotated images to it.
You can capture new images or add existing images:
```

---
linkTitle: "Create training dataset"

> title: "Create training dataset"

Author: I was originally trying to keep the line short enough to prevent it from wrapping. But I think this should fit!

description: "Create a dataset to train a machine learning model."
aliases:
  - /fleet/dataset/
  - /manage/data/label/
  - /manage/data/dataset/
  - /data/dataset/
  - /data-ai/ai/create-dataset/
---

To create a dataset:

{{< tabs >}}
{{% tab name="Web UI" %}}

1. Navigate to the **DATA** page and open the [**DATASETS** tab](https://app.viam.com/data/datasets).

1. Click the **+ Create dataset** button.

   {{< imgproc src="/services/data/create-dataset.png" alt="The **DATASET** tab of the **DATA** page, showing the **+ Create dataset** button." resize="800x" style="width:500px" class="imgzoom" >}}

1. Enter a unique name for the dataset.

1. Click **Create dataset** to create the dataset.

{{% /tab %}}
{{% tab name="CLI" %}}

1. First, install the Viam CLI and authenticate:

   {{< readfile "/static/include/how-to/install-cli.md" >}}

1. [Log in to the CLI](/dev/tools/cli/#authenticate).

1. Run the following command to create a dataset, replacing the `<org-id>` and `<name>` placeholders with your organization ID and a unique name for the dataset:

   ```sh {class="command-line" data-prompt="$"}
   viam dataset create --org-id=<org-id> --name=<name>
   ```

{{% /tab %}}

> {{% tab name="SDK" %}}

Author: Personally I like the UI/CLI/SDK top-level split because it limits the number of choices a user has to consider when they first reach a tabset. But I do agree that tabs-within-tabs are kind of awkward.

I'll try Web UI / CLI / Python SDK / Go SDK / Typescript SDK / Flutter SDK and see how that feels. If any of us feel strongly that we prefer it returned to the way it looks now, I'll change 'em back.

{{< tabs >}}
{{% tab name="Python" %}}

> To create a dataset, pass a unique dataset name and organization ID to `data_client.create_dataset`:

Reviewer 2: Yes good point - I think probably not optional, link can go on the `data_client.create_dataset`
